src/components/date/__internal__/date-formats/index.ts

```diff
@@ -140,7 +140,36 @@ interface LocaleFormats {
   format: string;
 }
 
-const getFormatData = ({ code = "en-GB" }: DateFnsLocale): LocaleFormats => {
+const getFormatData = (
+  { code = "en-GB" }: DateFnsLocale,
+  dateFormatOverride?: string
+): LocaleFormats => {
+  if (dateFormatOverride) {
+    const { format } = getOutputFormatForLocale(code);
+    let formatFromLocale;
+
+    switch (code) {
+      case "en-CA":
+      case "en-US":
+        formatFromLocale = "MM/dd/yyyy";
+        break;
+      case "ar-EG":
+      case "en-ZA":
+      case "fr-CA":
+        formatFromLocale = "dd/MM/yyyy";
+        break;
+      default:
+        formatFromLocale = format;
+    }
+
+    const formatsForLocale = getInputFormatsArrayForLocale(formatFromLocale);
+
+    return {
+      format: dateFormatOverride,
+      formats: generateFormats(formatsForLocale, "/"),
```

Comment:
**question**: I noticed we aren't generating the list of date formats from the override format string. @edleeks87 do these need to match by any chance?

```ts
return {
  format: dateFormatOverride,
  formats: generateFormats(
    getInputFormatsArrayForLocale(dateFormatOverride),
    "/"
  ),
}
```

Reply:
If they do need to match, what do you think about this tweak? This would help centralise our `date-fns` overrides:

```ts
const getFormatData = (
  { code = "en-GB" }: DateFnsLocale,
  dateFormatOverride?: string
): LocaleFormats => {

  // override settings of particular locales from `date-fns`
  const dateFnsFormatOverrides: Record<
    string,
    { format: string; separator?: Separator }
  > = {
    "en-CA": { format: "MM/dd/yyyy", separator: "/" },
    "en-US": { format: "MM/dd/yyyy", separator: "/" },
    "fr-CA": { format: "dd/MM/yyyy" },
    "en-ZA": { format: "dd/MM/yyyy" },
    "ar-EG": { format: "dd/MM/yyyy" },
  };

  const {
    format: defaultFormat,
    separator: defaultSeparator,
  } = getOutputFormatForLocale(code);

  const override = dateFnsFormatOverrides[code] || {};

  const format = dateFormatOverride ?? override.format ?? defaultFormat;

  const separator = override.separator ?? defaultSeparator;

  return {
    format,
    formats: generateFormats(
      getInputFormatsArrayForLocale(format),
      separator,
      getTrailingChar(format)
    ),
  };
};
```

Reply:
I _think_ as long as we're still able to support the same input variations and the output on blur formats to either the locale or the override format then we're okay

Reply:
I do think the suggested code is worth exploring at some point if not now as it's a bit tidier than the current implementation